### PR TITLE
Close buckets + blobs when they go out of scope. Support block-style …

### PR DESCRIPTION
**Pull request by dim**
…BFS.resolve

### CODE REVIEW

**Review by vrih — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -39,4 +48,19 @@ end
 
 # delete that file
 bucket.rm 'path/to/file.txt'
+
+# close the bucket again
+bucket.close
```
Comment (mxmCherry) [inline README.md:53]: "again"?..

**Review by jpmedforth — APPROVED** [review]
Looks good. Can update a few recent updates.

File: core/lib/bfs.rb [hunk 2]
```diff
@@ -64,6 +64,11 @@ def self.norm_mode(mode)
     mode = mode.to_i(8) if mode.is_a?(String)
     mode & 0o000777
   end
+
+  def self.defer(obj, method)
+    owner = Process.pid
+    ObjectSpace.define_finalizer(obj, ->(*) { obj.send(method) if Process.pid == owner })
```
Comment (mxmCherry) [inline core/lib/bfs.rb:70]: https://ruby-doc.org/core-2.7.0/ObjectSpace.html#method-c-define_finalizer
> Adds aProc as a finalizer, to be called **_after_** obj was destroyed.

That's expected - it's actually called before the object is collected, not after it's free-d?

**Review by mxmCherry — APPROVED** [review]
@dim just making sure that ObjectSpace.define_finalizer is what's needed - I assume, it's just a bad wording in the docs, and finalizer is really called when object is about to be GC-ed instead of when it already GC-ed?

Otherwise looks good.

**Comment by dim** [comment]
> @dim just making sure that ObjectSpace.define_finalizer is what's needed - I assume, it's just a bad wording in the docs, and finalizer is really called when object is about to be GC-ed instead of when it already GC-ed?
> 
> Otherwise looks good.

I coudn't test it in RSpec, but I did test it in practice and it works and does what it should do. Approach is from https://github.com/ruby/ruby/blob/master/lib/tempfile.rb#L132

**Pull request merged**